File: Assets/Scripts/Utils/AssetUtils.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using UnityEngine;

public class AssetUtils {
    public const string AppName = "";
    public const char AssetMapSplit = '|';
    public const string FileDetail = "__fileDetail";
    public const string AssetMap = "__assetmap";
    public const string VersionFileName = "version";

    /// <summary>
    /// 计算字符串的MD5值
    /// </summary>
    public static string GetStringMD5(string source)
    {
        MD5CryptoServiceProvider md5 = new MD5CryptoServiceProvider();
        byte[] data = System.Text.Encoding.UTF8.GetBytes(source);
        byte[] md5Data = md5.ComputeHash(data, 0, data.Length);
        md5.Clear();

        string destString = "";
        for (int i = 0; i < md5Data.Length; i++)
        {
            destString += System.Convert.ToString(md5Data[i], 16).PadLeft(2, '0');
        }
        destString = destString.PadLeft(32, '0');
        return destString;
    }

    /// <summary>
    /// 计算文件的MD5值
    /// </summary>
    public static string GetFileMD5ByPath(string file)
    {
        try
        {
            FileStream fs = new FileStream(file, FileMode.Open);
            System.Security.Cryptography.MD5 md5 = new System.Security.Cryptography.MD5CryptoServiceProvider();
            byte[] retVal = md5.ComputeHash(fs);
            fs.Close();

            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < retVal.Length; i++)
            {
                sb.Append(retVal[i].ToString("x2"));
            }
            return sb.ToString();
        }
        catch (Exception ex)
        {
            throw new Exception("md5file() fail, error:" + ex.Message);
        }
    }

    static string m_PersistentPath;
    public static string PersistentPath => m_PersistentPath;

    static string m_StreamAssetPath;
    public static string StreamAssetPath => m_StreamAssetPath;

    public static void LogBundleDtExist(string bundleName)
    {
        Debug.LogErrorFormat("AssetBundle what's bundleName = {0} don't exist", bundleName);
    }

    public static string GetPersistentFilePath(string file)
    {
        return string.Format("{0}/{1}", m_PersistentPath, file);
    }

    public static string GetStreamingFilePath(string file)
    {
#if UNITY_EDITOR
        // 使用bunle模式并且但是不走下载模式,那么本地服务器资源路径就是steaming路径
        if (AssetSimulate.USE_ASSET_BUNDLE && !AssetSimulate.USE_WEB_SERVER)
        {
            var simulateServer = GetWerServerBundlePath();
            return string.Format("{0}/{1}", simulateServer, file);
        }
#endif
        return string.Format("{0}/{1}", m_StreamAssetPath, file);
    }

    /// <summary>
    /// 获取可用文件路径 优先persistentPath 其次 steamingAssetsPath, 如果文件都不存在则返回null
    /// </summary>
    /// <param name="fileName"></param>
    /// <returns></returns>
    public static string GetValidFilePath(string fileName)
    {
        var persistentPath = AssetUtils.GetPersistentFilePath(fileName);
        if (File.Exists(persistentPath))
        {
            return persistentPath;
        }

        var streamingPath = AssetUtils.GetStreamingFilePath(fileName);
        if (File.Exists(streamingPath))
        {
            return streamingPath;
        }
        return null;
    }

    public static void LogAssetEmpty(string assetName, string type)
    {
        Debug.LogErrorFormat("nonexistent assetname = {0}  or  unconformable type(T) = {1} ", assetName, type);
    }

    static AssetUtils()
    {
        m_PersistentPath = Application.persistentDataPath;
        m_StreamAssetPath = Application.streamingAssetsPath;

        Debug.Log("m_PersistentPath = " + m_PersistentPath);
        Debug.Log("m_StreamAssetPath = " + m_StreamAssetPath);
    }

#if UNITY_EDITOR

    public static string GetWerServerBundlePath()
    {
        var pre = Application.dataPath.Remove(Application.dataPath.Length - 7);
        var plat = "android";
#if UNITY_IOS
        plat = "ios";
#endif
        return string.Format("{0}/Tools/WebServer/Assets/{1}", pre, plat);
    }


#endif
}

File: Assets/.lpproj/Apis/Unity/CS/AssetUtils.lua
---@class CS.AssetUtils : CS.System.Object
CS.AssetUtils = {}

---@field public CS.AssetUtils.AppName : CS.System.String
CS.AssetUtils.AppName = nil

---@field public CS.AssetUtils.AssetMapSplit : CS.System.Char
CS.AssetUtils.AssetMapSplit = nil

---@field public CS.AssetUtils.FileDetail : CS.System.String
CS.AssetUtils.FileDetail = nil

---@field public CS.AssetUtils.AssetMap : CS.System.String
CS.AssetUtils.AssetMap = nil

---@field public CS.AssetUtils.VersionFileName : CS.System.String
CS.AssetUtils.VersionFileName = nil

---@field public CS.AssetUtils.LuaFramework : CS.System.String
CS.AssetUtils.LuaFramework = nil

---@field public CS.AssetUtils.LuaScripts : CS.System.String
CS.AssetUtils.LuaScripts = nil

---@property readonly CS.AssetUtils.PersistentPath : CS.System.String
CS.AssetUtils.PersistentPath = nil

---@property readonly CS.AssetUtils.StreamAssetPath : CS.System.String
CS.AssetUtils.StreamAssetPath = nil

---@return CS.AssetUtils
function CS.AssetUtils()
end

---@param source : CS.System.String
---@return CS.System.String
function CS.AssetUtils.GetStringMD5(source)
end

---@param file : CS.System.String
---@return CS.System.String
function CS.AssetUtils.GetFileMD5ByPath(file)
end

---@param bundleName : CS.System.String
function CS.AssetUtils.LogBundleDtExist(bundleName)
end

---@param file : CS.System.String
---@return CS.System.String
function CS.AssetUtils.GetPersistentFilePath(file)
end

---@param file : CS.System.String
---@return CS.System.String
function CS.AssetUtils.GetStreamingFilePath(file)
end

---@param fileName : CS.System.String
---@return CS.System.String
function CS.AssetUtils.GetValidFilePath(fileName)
end

---@param assetName : CS.System.String
---@param type : CS.System.String
function CS.AssetUtils.LogAssetEmpty(assetName, type)
end

---@return CS.System.String
function CS.AssetUtils.GetWerServerBundlePath()
end